encodings/alp/src/alp/mod.rs

```diff
@@ -195,79 +239,6 @@ pub trait ALPFloat: private::Sealed + Float + Display + 'static {
     }
 }
 
-#[allow(clippy::cast_possible_truncation)]
-fn encode_chunk_unchecked<T: ALPFloat>(
-    chunk: &[T],
-    exp: Exponents,
-    encoded_output: &mut BufferMut<T::ALPInt>,
-    patch_indices: &mut BufferMut<u64>,
-    patch_values: &mut BufferMut<T>,
-    fill_value: &mut Option<T::ALPInt>,
-) {
-    let num_prev_encoded = encoded_output.len();
-    let num_prev_patches = patch_indices.len();
-    assert_eq!(patch_indices.len(), patch_values.len());
-    let has_filled = fill_value.is_some();
-
-    // encode the chunk, counting the number of patches
-    let mut chunk_patch_count = 0;
-    encoded_output.extend(chunk.iter().map(|v| {
-        let encoded = unsafe { T::encode_single_unchecked(*v, exp) };
-        let decoded = T::decode_single(encoded, exp);
-        let neq = (decoded != *v) as usize;
-        chunk_patch_count += neq;
-        encoded
-    }));
```

Comment:
This loop structure is actually pretty important for performance.

You can refer to the original code where this is actually separated out into two loops:


Loop 1: calculate and materialize encoded + decoded vectors

Loop 2: find the exceptions by looping through the zipped vectors to find mismatches

https://github.com/cwida/ALP/blob/main/include/alp/encoder.hpp#L338-L346

https://github.com/cwida/ALP/blob/main/include/alp/encoder.hpp#L370-L376

I'm not sure i personally find the new version on the right more readable, but I want to make sure we keep a similar structure that the branch predictor likes